src/i_midimusic.c

```diff
@@ -1854,53 +1566,28 @@ static void I_WIN_UnRegisterSong(void *handle)
     song.rpg_loop = false;
 }
 
-static void I_WIN_ShutdownMusic(void)
+static void I_MID_ShutdownMusic(void)
 {
-    MMRESULT mmr;
-
-    if (!hMidiStream)
+    if (!music_initialized)
     {
         return;
     }
 
-    EnterCriticalSection(&CriticalSection);
-    StreamStop();
-    I_WIN_UnRegisterSong(NULL);
-    win_midi_state = STATE_SHUTDOWN;
-    StreamStart();
-    LeaveCriticalSection(&CriticalSection);
+    I_MID_StopSong(NULL);
+    I_MID_UnRegisterSong(NULL);
 
-    if (WaitForSingleObject(hPlayerThread, PLAYER_THREAD_WAIT_TIME) == WAIT_OBJECT_0)
-    {
-        CloseHandle(hPlayerThread);
-        hPlayerThread = NULL;
-    }
-    StreamStop();
+    // Send notes/sound off prior to reset to prevent volume spikes.
+    SendNotesSoundOff();
+    ResetDevice();
 
-    if (buffer.prepared)
-    {
-        UnprepareHeader();
-        buffer.prepared = false;
-    }
-
-    mmr = midiStreamClose(hMidiStream);
-    if (mmr != MMSYSERR_NOERROR)
-    {
-        MidiError("midiStreamClose", mmr);
-    }
-    hMidiStream = NULL;
+    rtmidi_close_port(midiout);
```

Comment:
There needs to be a check before calling this function to ensure that any pending events are clear from the given OS's midi system. On Windows (not sure about others), it's possible to drop portions of the pitch bend and volume reset messages that are called in `ResetDevice()` because `rtmidi_close_port()` simply calls WinMM's `midiOutClose()` blindly.

Reply:
For comparison, before this PR, we carefully waited for `STATE_SHUTDOWN` -> `WaitForSingleObject` (confirms message buffer has been returned) -> `STATE_EXIT` -> reset of shutdown process. Not sure how it would work now.

Reply:
I think we can just add a little delay.

Reply:
Now we check `MIDIERR_STILLPLAYING`, see WINMM `MIDI_SendLongMsg` in midiout.c